src/daft-local-execution/src/intermediate_ops/inner_hash_join_probe.rs

```diff
@@ -0,0 +1,181 @@
+use std::sync::Arc;
+
+use common_error::DaftResult;
+use daft_core::prelude::SchemaRef;
+use daft_dsl::ExprRef;
+use daft_micropartition::MicroPartition;
+use daft_table::{GrowableTable, Probeable, Table};
+use indexmap::IndexSet;
+use tracing::{info_span, instrument};
+
+use super::intermediate_op::{
+    IntermediateOperator, IntermediateOperatorResult, IntermediateOperatorState,
+};
+use crate::pipeline::PipelineResultType;
+
+enum InnerHashJoinProbeState {
+    Building,
+    ReadyToProbe(Arc<dyn Probeable>, Arc<Vec<Table>>),
+}
+
+impl InnerHashJoinProbeState {
+    fn set_table(&mut self, table: &Arc<dyn Probeable>, tables: &Arc<Vec<Table>>) {
+        if let InnerHashJoinProbeState::Building = self {
+            *self = InnerHashJoinProbeState::ReadyToProbe(table.clone(), tables.clone());
+        } else {
+            panic!("InnerHashJoinProbeState should only be in Building state when setting table")
+        }
+    }
+
+    fn get_probeable_and_table(&self) -> (&Arc<dyn Probeable>, &Arc<Vec<Table>>) {
+        if let InnerHashJoinProbeState::ReadyToProbe(probe_table, tables) = self {
+            (probe_table, tables)
+        } else {
+            panic!("get_probeable_and_table can only be used during the ReadyToProbe Phase")
+        }
+    }
+}
+
+impl IntermediateOperatorState for InnerHashJoinProbeState {
+    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
+        self
+    }
+}
+
+pub struct InnerHashJoinProbeOperator {
+    probe_on: Vec<ExprRef>,
+    common_join_keys: Vec<String>,
+    left_non_join_columns: Vec<String>,
+    right_non_join_columns: Vec<String>,
+    build_on_left: bool,
+}
+
+impl InnerHashJoinProbeOperator {
+    pub fn new(
+        probe_on: Vec<ExprRef>,
+        left_schema: &SchemaRef,
+        right_schema: &SchemaRef,
+        build_on_left: bool,
+        common_join_keys: IndexSet<String>,
+    ) -> Self {
+        let left_non_join_columns = left_schema
+            .fields
+            .keys()
+            .filter(|c| !common_join_keys.contains(*c))
+            .cloned()
+            .collect();
+        let right_non_join_columns = right_schema
+            .fields
+            .keys()
+            .filter(|c| !common_join_keys.contains(*c))
+            .cloned()
+            .collect();
+        let common_join_keys = common_join_keys.into_iter().collect();
+        Self {
+            probe_on,
+            common_join_keys,
+            left_non_join_columns,
+            right_non_join_columns,
+            build_on_left,
+        }
+    }
+
+    fn probe_inner(
+        &self,
+        input: &Arc<MicroPartition>,
+        state: &mut InnerHashJoinProbeState,
+    ) -> DaftResult<Arc<MicroPartition>> {
+        let (probe_table, tables) = state.get_probeable_and_table();
```

Comment:
This may be worth making a struct type for

```
struct ProbeState {
  probe_table
  tables
}
```

Reply:
Makes sense! Implemented it